internal/configs/configurator.go

```diff
@@ -454,6 +452,70 @@ func (cnf *Configurator) AddOrUpdateVirtualServers(virtualServerExes []*VirtualS
 	return allWarnings, nil
 }
 
+func (cnf *Configurator) updateTransportServerMetricsLabels(transportServerEx *TransportServerEx, upstreams []version2.StreamUpstream) {
+	labels := make(map[string][]string)
+	newUpstreams := make(map[string]bool)
+	var newUpstreamsNames []string
+
+	upstreamServerPeerLabels := make(map[string][]string)
+	newPeers := make(map[string]bool)
+	var newPeersIPs []string
+
+	for _, u := range upstreams {
+		labels[u.Name] = []string{u.UpstreamLabels.Service, u.UpstreamLabels.ResourceType, u.UpstreamLabels.ResourceName, u.UpstreamLabels.ResourceNamespace}
+		newUpstreams[u.Name] = true
+		newUpstreamsNames = append(newUpstreamsNames, u.Name)
+
+		for _, server := range u.Servers {
+			podName := transportServerEx.PodsByIP[server.Address]
+			labelKey := fmt.Sprintf("%v/%v", u.Name, server.Address)
+			upstreamServerPeerLabels[labelKey] = []string{podName}
+
+			newPeers[labelKey] = true
+			newPeersIPs = append(newPeersIPs, labelKey)
+		}
+	}
+
+	key := fmt.Sprintf("%v/%v", transportServerEx.TransportServer.Namespace, transportServerEx.TransportServer.Name)
+
+	removedPeers := findRemovedKeys(cnf.metricLabelsIndex.transportServerUpstreamPeers[key], newPeers)
+	cnf.metricLabelsIndex.transportServerUpstreamPeers[key] = newPeersIPs
+
+	removedUpstreams := findRemovedKeys(cnf.metricLabelsIndex.transportServerUpstreams[key], newUpstreams)
+	cnf.metricLabelsIndex.transportServerUpstreams[key] = newUpstreamsNames
+
+	if cnf.isPlus {
+		cnf.labelUpdater.UpdateStreamUpstreamServerPeerLabels(upstreamServerPeerLabels)
+		cnf.labelUpdater.DeleteStreamUpstreamServerPeerLabels(removedPeers)
+		cnf.labelUpdater.UpdateStreamUpstreamServerLabels(labels)
+		cnf.labelUpdater.DeleteStreamUpstreamServerLabels(removedUpstreams)
+
+		streamServerZoneLabels := make(map[string][]string)
+		newZones := make(map[string]bool)
+		newZonesNames := []string{transportServerEx.TransportServer.Spec.Host}
+
+		streamServerZoneLabels[transportServerEx.TransportServer.Spec.Host] = []string{
+			"transportserver", transportServerEx.TransportServer.Name, transportServerEx.TransportServer.Namespace}
+
+		newZones[transportServerEx.TransportServer.Spec.Host] = true
```

Comment:
Im confused @pleshakov , 
`transportServerEx.TransportServer.Spec.Host` will not be empty for non-TLSPassthrough transport servers? Is [this](https://github.com/nginxinc/kubernetes-ingress/blob/annotate-tcp/internal/configs/configurator_test.go#L1008) example TS wrong so? and I would've thought it would be the other way around. Using `transporterServerEx.TransportServer.Spec.Listener.Name` would be bad as it would be empty for non-TLSPassthrough TS's.

Reply:
Also for both [VS](https://github.com/nginxinc/kubernetes-ingress/blob/annotate-tcp/internal/configs/configurator.go#L358) and [Ingress](https://github.com/nginxinc/kubernetes-ingress/blob/annotate-tcp/internal/configs/configurator.go#L202) we use the host name as the map key, is there a reason to not follow that for TransportServers?

Reply:
Please take a look at the examples:
- TLSPassthrogh transport server -- https://github.com/nginxinc/kubernetes-ingress/blob/master/examples-of-custom-resources/tls-passthrough/transport-server-passthrough.yaml
- TCP and UDP transport servers -- https://github.com/nginxinc/kubernetes-ingress/tree/master/examples-of-custom-resources/basic-tcp-udp

> Also for both VS and Ingress we use the host name as the map key, is there a reason to not follow that for TransportServers?

yep. there is no host for TCP/UDP load balancing